### ISSUE TITLE
Why block requests for missing resources?

### ISSUE THREAD

**Issue by d0mf1g** [Synzvato/decentraleyes#239]
https://github.com/Synzvato/decentraleyes/wiki/Frequently-Asked-Questions#why-doesnt-it-deliver-resources-from-cdns-i-block-using-a-different-add-on

From my understanding (correct me if I'm wrong), strict blockers block all CDNs from loading, meaning there are no CDN requests for Decentraleyes to intercept and do local injections. Whitelisting the CDNs allows the requests to go through and thus allows Decentraleyes to look for the local files first.

But why are we blocking requests  for missing resources from Decentraleyes? Because the strict blocker add on is blocking them anyway (outside of the whitelist), and so it makes no sense for Decentraleyes to keep looking for CDNs?

**Comment by Synzvato** [comment]
> But why are we blocking requests for missing resources [...]? Because the [...] blocker [...] is blocking them anyway [...], and so it makes no sense for Decentraleyes to keep looking for CDNs?

Hi @d0mf1g, and thanks for sharing your concerns! As soon as you apply the proposed rules, your regular content blocker will no longer interfere with requests that are interceptable by Decentraleyes. By enabling ```Block requests for missing resources```, you remain in control of the outbound traffic.

I hope this answers your question. If not, please let me know!

**Comment by d0mf1g** [comment]
Why would blocking outbound connections be important in the case of using Decentraleyes with another blocker? 

And what wouldn't pages break if local resources are not available and outbound connections for the CDN are blocked?

**Comment by Synzvato** [comment]
> Why would blocking [...] be important in the case of using Decentraleyes with another blocker?

The rules you add, tell the blocker to ignore any requests that are supported by Decentraleyes. This means that they are delegated, and that Decentraleyes needs to figure out what to do with them. When Decentraleyes is forced to allow a delegated request, your regular content blocker will not interfere.

> [...] wouldn't pages break if local resources are not available and [...] connections [...] are blocked?

That is correct. Do not enable ```Block requests for missing resources``` if you expect websites to work without any manual intervention. Only follow the guide if you are familiar with managing strict rules.

**Comment by d0mf1g** [comment]
Sorry, not sure I completely follow. An example would help. What type of interference would there be if I left the "block requests for missing resources" unchecked while my content blocker is active?

Thanks.

**Comment by Synzvato** [comment]
That depends on whether or not you have applied the proposed rules. Once you add them, your content blocker will not be touching any supported CDN-requests, _even_ if your blocker is otherwise enabled.

So, if you leave ```Block requests for missing resources``` unchecked and apply the rules, the blocker refuses to touch any supported ```CDN```-requests, and Decentraleyes only prevents requests for resources that are locally available. This means that requests for missing resources _will_ be sent off to a ```CDN```.

Here are some examples to illustrate what happens when a file isn't locally available:

```[X] REGULAR BLOCKER``` ```->``` ```[X] DECENTRALEYES``` ```->``` ```[ ] CDN```
> _**Causes** breakage, **non-strict** blocking policies, and block missing **enabled**._

```[X] REGULAR BLOCKER``` ```->``` ```[X] DECENTRALEYES``` ```->``` ```[X] CDN```
> _**Does not cause** breakage, **non-strict** blocking policies, and block missing **disabled**._

```[X] REGULAR BLOCKER``` ```->``` ```[ ] DECENTRALEYES``` ```->``` ```[ ] CDN```
> _**Causes** breakage, **strict** blocking policies, and block missing **disabled**._

**Comment by d0mf1g** [comment]
Following so far.

>This means that requests for missing resources will be sent off to a CDN

But why would that be an issue for someone using a content blocker? The FAQ makes it look like disabling requests for missing resources has to be done in conjunction with using a content blocker.

**Comment by Synzvato** [comment]
This is not an issue, unless you have applied any strict blocking policies. The FAQ does convey this, and also explains that regular users do not need to follow the guide. From the corresponding answer:

> Easy mode is uBlock Origin's default mode. Unless you have manually applied a more strict blocking mode (e.g. medium mode), you do not need to follow the steps below.

**Comment by d0mf1g** [comment]
Yes, I was referring to strict blocking. Why would sending requests for missing resources be an issue with, say something like uBlock advanced mode? Why does the FAQ tell us to disable requests for missing resources when using strict blocking?

When using a strict blocker, why are we doing this:

[X] REGULAR BLOCKER -> [X] DECENTRALEYES -> [ ] CDN
>block missing enabled.

and not this:

[X] REGULAR BLOCKER -> [X] DECENTRALEYES -> [X] CDN
>block missing disabled.

**Comment by Synzvato** [comment]
@d0mf1g Because when you add the rules from the guide, your regular blocker becomes less strict and will no longer block any outgoing CDN requests that are interceptable by Decentraleyes.

**Comment by d0mf1g** [comment]
I understand the whitelisting part, but why change the settings in Decentraleyes to block missing resources?

From the FAQ: 

>Important: Enable the block requests for missing resources setting inside the Decentraleyes extension Options screen, in order to reinstate your strict CDN-blocking policies.

It says Decentraleyes respects the blocker (blocker takes priority). If domains to missing resources are already blocked by uBlock (again for example), isn't this step redundant?

**Comment by Synzvato** [comment]
> It says Decentraleyes respects the blocker (blocker takes priority).

@d0mf1g Decentraleyes will not inject resources that are blocked by other content blockers. This is why the guide asks you to instruct your strict blockers to _completely ignore_ any supported CDN-requests.

> I understand the whitelisting part, but why change the settings [...]? If domains to missing resources are already blocked by uBlock (again for example), isn't this step redundant?

Once you apply these whitelist rules, such domains are _no longer_ blocked by e.g. uBlock Origin. In other words, other blockers will no longer interfere, and will delegate _all CDN-requests_ to Decentraleyes.

So no, the step is not redundant. It prevents traffic to domains that are no longer blacklisted.

**Comment by d0mf1g** [comment]
>Once you apply these whitelist rules, such domains are no longer blocked by e.g. uBlock Origin. In other words, other blockers will no longer interfere, and will delegate all CDN-requests to Decentraleyes.

>So no, the step is not redundant. It prevents traffic to domains that are no longer blacklisted.

Ohhh ok. It makes sense to me now. Basically it's to stop requests for missing resources from even whitelisted domains. Of course there would be cases where resources are missing from the whitelisted domains, and so Decentraleyes would then communicate to those domains to get the missing resources. 

This whole time I was under the assumption that Decentraleyes would already have all of the resources from the whitelisted domains. I hadn't thought about missing resources from whitelisted domains. 

Thanks for taking the time to clarify this issue for me. Can I just suggest some changes to this FAQ?  Just to expound the answer. Changes in bold.

>Most content blockers do not block delivery networks by default, as doing so breaks pages. This extension works out of the box, unless it's asked to operate under strict blocking rules. Any policies set by other extensions are respected**; if delivery networks are blocked by a content blocker, then there would be no requests for Decentraleyes to intercept, and thus locally available resources will not be served through Decentraleyes. A workaround would be to add the delivery networks' domains to the content blocker's whitelist so that Decentraleyes can then intercept the now allowed requests and load the resources locally, if they are available.**

>Important:  **In order to prevent outbound connections to even whitelisted content delivery networks, you can enable** the _block requests for missing resources_ setting inside the Decentraleyes extension Options screen. **This will force Decentraleyes to load only local resources, if they are available. Requests for missing resources from whitelisted domains will not be made and may cause webpages to break.**

What do you think?

**Comment by Synzvato** [comment]
> [...] I was under the assumption that Decentraleyes would [...] have all of the resources [...]. I hadn't thought about missing resources [...]. Thanks for taking the time to clarify this issue for me.

You're welcome! I'm very happy to hear that the explanations were helpful.

> Can I just suggest some changes to this FAQ?

While I understand that you'd like things to be a litle bit more explicit, I would like to keep the current instructions as concise as possible. From my experience, longer paragraphs mean that less people will properly read their contents. I have [further clarified](https://github.com/Synzvato/decentraleyes/wiki/Frequently-Asked-Questions/_compare/53bf3194ccb623a4d37be44bee51f72578294081...647e7a0abe326815996155b0b93e8b639dd55d4f) that this is an advanced feature, though.

**Comment by User486375** [comment]
Thought it would be a good idea for me to respond here rather than creating a new issue. 

I recently enabled this setting along with the noop rules for **medium mode** blocking on uBlock origin and I noticed a good few sites breaking and me having to white list them with Decentraleyes. Why are they blocked but not locally available?

**Comment by vertigo220** [comment]
I'm with @User486375 here. I just recently set this up, and have found some sites not working, which requires me to disable Decentraleyes on them. It seems that by enabling the "Block requests for missing resources" option, you're risking breaking sites and having to open them up completely, eliminating all protection on those sites, with no real advantage. If you just leave that option disabled, then any missing resources (but _only_ the missing ones) would be requested, but Decentraleyes would strip identifying info from the requests, which it wouldn't be doing if you had disabled it for the site to get the site to work. So what am I missing here? Why break sites, requiring manual intervention (not to mention it usually takes a while to figure out, since uBo and uMatrix are typically the first suspects) and _decreased_ protection in those cases? After all, the only time it would be blocking something is when you do get breakage, so the only time it would be "protecting" you it would actually have the opposite effect, since in those instances (assuming you want to access the site) you'll have to disabled Decentraleyes completely. Is that logic sound or is there more to it?

**Comment by ArchangeGabriel** [comment]
Blocking is not the default. You should use it only if you want no connection ever to any CDNs. Else, I’d suggest to keep it disabled.

**Comment by ArchangeGabriel** [comment]
Decentraleyes can only strip “some” identifying bits, not all. Your IP address for instance will stay. So yes, it does matter for some people.

Yes, CDN are missing, and until they are added, you are less protected indeed.
